x-pack/plugins/discover_log_explorer/public/state_machines/datasets/src/types.ts

```diff
@@ -0,0 +1,91 @@
+/*
+ * Copyright Elasticsearch B.V. and/or licensed to Elasticsearch B.V. under one
+ * or more contributor license agreements. Licensed under the Elastic License
+ * 2.0; you may not use this file except in compliance with the Elastic License
+ * 2.0.
+ */
+import { DoneInvokeEvent } from 'xstate';
+import type { IHashedCache } from '../../../../common/hashed_cache';
+import { FindDatasetsResponse, SortOrder } from '../../../../common/latest';
+import { Dataset } from '../../../../common/datasets';
+
+export interface DatasetsSearchParams {
+  datasetQuery?: string;
+  sortOrder?: SortOrder;
+}
+
+export interface WithCache {
+  cache: IHashedCache<DatasetsSearchParams, FindDatasetsResponse>;
+}
+
+export interface WithSearch {
+  search: DatasetsSearchParams;
+}
+
+export interface WithDatasets {
+  datasets: Dataset[];
+}
+
+export interface WithNullishDatasets {
+  datasets: null;
+}
+
+export interface WithError {
+  error: Error;
+}
+
+export interface WithNullishError {
+  error: null;
+}
+
+export type DefaultDatasetsContext = WithCache &
+  WithNullishDatasets &
+  WithSearch &
+  WithNullishError;
+
+type LoadingDatasetsContext = DefaultDatasetsContext;
+
+type LoadedDatasetsContext = WithCache & WithDatasets & WithSearch & WithNullishError;
+
+type LoadingFailedDatasetsContext = WithCache & WithNullishDatasets & WithSearch & WithError;
+
+export type DatasetsTypestate =
+  | {
+      value: 'uninitialized';
+      context: DefaultDatasetsContext;
+    }
+  | {
+      value: 'loading';
+      context: LoadingDatasetsContext;
+    }
+  | {
+      value: 'loaded';
+      context: LoadedDatasetsContext;
+    }
+  | {
+      value: 'loadingFailed';
+      context: LoadingFailedDatasetsContext;
+    }
+  | {
+      value: 'debounceSearchingDatasets';
+      context: LoadedDatasetsContext;
+    };
+
+export type DatasetsContext = DatasetsTypestate['context'];
+
+export type DatasetsEvent =
+  | {
+      type: 'LOAD_DATASETS';
+    }
+  | {
+      type: 'RELOAD_DATASETS';
+    }
+  | {
+      type: 'SEARCH_DATASETS';
+      search: DatasetsSearchParams;
+    }
+  | {
+      type: 'SORT_DATASETS';
+      search: DatasetsSearchParams;
+    }
+  | DoneInvokeEvent<any>;
```

Comment:
Sure, I read a lot of discussions around this topic on the state repo, and since wasn't possible to automatically infer the type for these DoneEvent (unless the action was inline in the machine), the proposed workarounds were 2:
- Use `DoneInvokeEvent<any>;`
- Type the DoneInvokeEvent generic and add guards to the actions that use it.

I updated now the typing to `DoneInvokeEvent<FindDatasetValue | Error>` and I applied the necessary guards, it should now be typed safely (I did the same for the integrations machine) 👍